src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md

```diff
@@ -0,0 +1,862 @@
+
+    public Point() {}
+
+    public Point(int x, int y) => (X, Y) = (x, y);
+}
+```
+
+The public parameterless constructor is used:
+
+```C#
+Point point = JsonSerializer.Deserialize<Point>(@"{""X"":1,""Y"":2}");
+Console.WriteLine(point.X); // 0
+Console.WriteLine(point.Y); // 0
+```
+
+##### `struct`s will always be deserialized using the default constructor if `[JsonConstructor]` is not used
+
+Given `Point`,
+
+```C#
+public struct Point
+{
+    public int X { get; }
+
+    public int Y { get; }
+
+    public Point(int x, int y) => (X, Y) = (x, y);
+}
+```
+
+The default constructor is used:
+
+```C#
+Point point = JsonSerializer.Deserialize<Point>(@"{""X"":1,""Y"":2}");
+Console.WriteLine(point.X); // 0
+Console.WriteLine(point.Y); // 0
+```
+
+##### A single public parameterized constructor will always be used if there's no public parameterless constructor
+
+Given `Point`,
+
+```C#
+public class Point
+{
+    public int X { get; }
+
+    public int Y { get; }
+
+    public Point(int x, int y) => (X, Y) = (x, y);
+}
+```
+
+The singular parameterized constructor is used:
+
+```C#
+Point point = JsonSerializer.Deserialize<Point>(@"{""X"":1,""Y"":2}");
+Console.WriteLine(point.X); // 1
+Console.WriteLine(point.Y); // 2
+```
+
+This rule does not apply to `struct`s as there's always a public parameterless constructor.
+
+##### `NotSupportedException` is thrown when there are multiple parameterized ctors, but no public parameterless ctor
+
+Given another definition for `Point`,
+
+```C#
+public class Point
+{
+    public int X { get; }
+
+    public int Y { get; }
+
+    public int Z { get; }
+
+    public Point(int x, int y) => (X, Y) = (x, y);
+
+    public Point(int x, int y, int z = 3) => (X, Y, Z) = (x, y, z);
+}
+```
+
+A `NotSupportedException` is thrown because it is not clear which constructor to use. This may be resolved by using the `[JsonConstructor]`, or by adding a
+public parameterless constructor.
+
+```C#
+Point point = JsonSerializer.Deserialize<Point>(@"{""X"":1,""Y"":2,""Z"":3}"); // Throws `NotSupportedException`
+```
+
+This rule does not apply to `struct`s as there's always a public parameterless constructor.
+
+#### Using [JsonConstructor]
+
+##### Non-`public` constructors will not be used unless specified with a `[JsonConstructor]` attribute
+
+The serializer doesn't offer non-`public` support by default. An attribute should be placed on any non-`public` constructor
+that is desired to be used during deserialization, even if they are parameterless.
+Ideally, people should not be able to deserialize instances of types they don't own using non-`public` constructors.
+This restriction of an attribute based opt-in, rather than a globally applied opt-in, dissuades the violation of constructor access modifiers.
+
+Given `Point`:
+
+```C#
+public class Point
+{
+    public int X { get; }
+
+    public int Y { get; }
+
+    public int Z { get; }
+
+    [JsonConstructor]
+    private Point(int x, int y) => (X, Y) = (x, y);
```

Comment:
I'm a bit concerned about features like this because most people (and that includes us) don't really reason about privates as external state. I'd sleep better if we could make this `protected`, for instance.

Reply:
Agree. We should verify that we really, really need non-public support.

`private` and `protected` support won't support most code-gen designs -- the converter and POCO must be one and the same. `internal` is a little more palatable.

Reply:
Depends on how we do the source gen. Wasn't one proposal that we'd put them on the same type?

Reply:
> We should verify that we really, really need non-public support

Agreed. We have a similar concern with the upcoming non-public getter/setter support (https://github.com/dotnet/runtime/issues/29743) feature, but we are committed to doing this.

Reply:
I think if folks realize what they are doing it might be ok.  Something like this:
![image](https://user-images.githubusercontent.com/8918108/76580316-bdc19180-648c-11ea-82dc-1bbb6cab7d6b.png)

Reply:
The resolution is to wait for an explicit customer ask before adding this.